### ISSUE TITLE
LCD: draw points/lines/ and other geometrical forms

### ISSUE THREAD

**Issue by radut** [Pi4J/pi4j-v1#84]
Hi, 
What about improving the pi4j lcd library with the posibility to draw points, lines and base geometric forms ?

**Comment by SrMouraSilva** [comment]
Lcd library uses WiringPi Lcd Library (http://wiringpi.com/dev-lib/lcd-library/).
I dont know if are possible changes at points level

**Comment by SrMouraSilva** [comment]
Yes, it's possible

https://github.com/noxo/SnakePI4J
https://github.com/noxo/SnakePI4J/blob/master/src/org/noxo/devices/driver/JPCD8544.java

**Comment by natdan** [comment]
LOL - you got me with 'awt' part of your post! but seriously, I don't think it would be too hard to have Swing draw on any framebuffer. It is not 'implementation' (of current OS) specific and theoretically easy to be used anywhere...

**Comment by SrMouraSilva** [comment]
"LOL - you got me with 'awt' part of your post!" I'm sorry! 

But...

"THE EMBEDDED WINDOW TOOLKIT
Porting Java’s AWT to an Embedded System"

http://alumni.soe.ucsc.edu/~wholt/thesis.pdf

**Comment by SrMouraSilva** [comment]
I do not know, but I think Pi4j do not support Displays graphics

João Sombra has [ported](https://github.com/CleversonSA/JPCD8544) a python port of a [C implementation of PCD8544](https://github.com/binerry/RaspberryPi/tree/master/libraries/c/PCD8544) (of a arduino implementation...)

(Reference: https://www.youtube.com/watch?v=LLt9q1CAxRs)

The C implementation has a comment in your [github project](https://github.com/binerry/RaspberryPi/tree/master/libraries/c/PCD8544)

"This driver uses 5 GPIOs on target board with a bit-bang SPI implementation (hence, may not be as fast but seems fast enough)."

João Sombra answered my question in your [demo video](https://www.youtube.com/watch?v=LLt9q1CAxRs)
"Funcionou mas o desempenho ficou bem a desejar." 
(Google translated: It worked but the performance was much to be desired.)

I need work with PCD8544, then I will test. If has a success, I will pull request then

**Comment by SrMouraSilva** [comment]
Exists a cool c++ lib with PCD8544 support: [Rpi-hw](https://github.com/Wicker25/Rpi-hw/)

I'm putting some links in here for future reference
- Code to port: https://github.com/Wicker25/Rpi-hw/blob/master/src/display/pcd8544.cpp
- Display in Raspberry 2 support: Wicker25/Rpi-hw#15 https://github.com/Wicker25/Rpi-hw/issues/15

**Comment by SrMouraSilva** [comment]
@natdan, I added [a support for Java Graphics](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/drawer/DisplayGraphics.java) for [display](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/Display.java)

If AWT uses Graphics2D, maybe it's possible portable it for display

I added a [initial support for PCD8544](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/impl/PCD8544DisplayComponent.java) (Nokia 5110) (not based in I2C data communication

[Test example](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java), it's possible:
- [Write text](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java#L48)
- [Set the font](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java#L52-L53)
- [Draw image](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java#L64-L66)
- [Draw lines](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java#L78)
- [Draw rects](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java#L92)
- [Fill rects](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java#L105)
- [Draw circle](https://github.com/SrMouraSilva/JavaPedalMIDI/blob/master/Display/src/com/pi4j/component/display/test/Main.java#L116)
- (Maybe) all of [graphics](https://docs.oracle.com/javase/7/docs/api/java/awt/Graphics.html) features (i'm not tested all)

In future, port [Display2D](https://docs.oracle.com/javase/7/docs/api/java/awt/Graphics2D.html) also  is a good idea, I think

**Comment by natdan** [comment]
Yes! That is the way to do it! Graphics (Graphics2D) is class to implement and everything else should use it! Now - do you want to contribute your impl. to pi4j project, maybe? Looking at your code I think we should have whole section for displays over here.

**Comment by SrMouraSilva** [comment]
@natdan , it's my clain :smile:

A rapid test
https://www.youtube.com/watch?v=olnMZMNEJqk

I will implement the Graphics2D and forked the... What's the branch?

(I'm sorry by my bad english :no_mouth:)

**Comment by natdan** [comment]
Let me know when you fork it. I've just procured i2c 128x64 bw display and I am happy to help - give you a hand. So if/when you fork it send me a pm or write here if you want some help.

**Comment by SrMouraSilva** [comment]
@natdan I'm preparing the code for fork and pull request :smiley: 
https://github.com/SrMouraSilva/JavaPedalMIDI/commit/21b58e22ec0a3a698643c6376711f37ae017fc07

**Comment by natdan** [comment]
Thanks. BTW Forgot to say - it looks great! :)

**Comment by SrMouraSilva** [comment]
:)

**Comment by SrMouraSilva** [comment]
@savageautomate, please add RELEASE 2.0 to this because my pull request #84 use Java 8
